docs/extending/extending_client_side.md

```diff
@@ -0,0 +1,531 @@
+
+The `w-progress` Controller is used to help the user avoid clicking the same button multiple times when there may be some delayed behaviour required. Custom usage of this is supported for usage within the admin interface.
+
+-   `<button ... data-w-progress-duration-value="500" ...>` - custom duration can be declared on the element
+-   `<button ... class="custom-button" data-w-progress-active-class="custom-button--busy" ...>` - custom 'active' class to replace the default `button-longrunning-active` (must be a single string without spaces)
+-   `<button ... ><strong data-w-progress-target="label">{% trans 'Create' %}</strong></button>` - any element can be the button label (not just `em`)
+-   `<button ... data-action="w-progress#activate focus->w-progress#activate" ...>` - any event can be used to trigger the in progress behaviour
+-   `<button ... data-action="w-progress#activate:once" ...>` - only trigger the progress behaviour once
+-   `<button ... data-action="readystatechange@document->w-progress#activate:once" data-w-progress-duration-value="5000" disabled ...>` - disabled on load (once JS starts) and becomes enabled after 5s duration
+
+### Debugging & error handling
+
+For simple debugging, you can enable the [Stimulus debug mode](https://stimulus.hotwired.dev/handbook/installing#debugging).
+
+```javascript
+window.Stimulus.debug = true;
+```
+
+You can use also the built in [Stimulus error callback](https://stimulus.hotwired.dev/handbook/installing#error-handling) for more robust error handling.
+
+```javascript
+window.onerror = console.error;
+```
+
+### Testing your Controllers
+
+Wagtail uses an out of the box Stimulus application instance, you should be able to write your tests with your own package.json installation of Stimulus without issue.
+You should not need to mock the `Stimulus.use.Controller`, it is best you write your Controllers as extending `import {Controller} from '@hotwired/stimulus';` and then use any build tooling to redirect this to the global in production builds.
+
+### Advanced Stimulus overrides
+
+While these kinds of overrides are supported as a last ditch method to fully customise behaviour, writing this knd of code will require you to understand the existing implementations and support the JavaScript on your own. It is also important to note that the maintenance burden will be on your project to ensure that any functionality added or changes to controller identifiers are supported.
+
+If you find yourself reaching for this to fix a bug, please ensure an issue is raised on the Wagtail repository with your functional work around.
+
+Similarly, if some common Stimulus Controller usage would be helpful for the community, such as additional values, targets, DOM events or methods, please raise an issue with the suggestion.
+
+```{warning}
+Usage of `class ... extends` from the globally provided `window.Stimulus.use.Controller` cannot be done outside of code that is also bundled to ES5.
+```
```

Comment:
This seems to be repeated in the document, is it intentional?

Reply:
Thanks, this line was actually incorrect (before we moved to ES6+ targets), so I have removed it.